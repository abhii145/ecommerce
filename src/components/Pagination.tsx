import React, { useCallback, useMemo } from "react"
import { FaAngleLeft, FaAngleRight } from "react-icons/fa"
import { PaginationProps } from "../types"

const Pagination: React.FC<PaginationProps> = ({
  currentPage,
  total,
  limit,
  onPageChange,
}) => {
  const totalPages = useMemo(() => Math.ceil(total / limit), [total, limit])

  const handlePageChange = useCallback((page: number) => {
    if (page >= 1 && page <= totalPages) {
      onPageChange(page)
    }
  }, [totalPages, onPageChange])

  return (
    <div className="flex items-center justify-between border-t border-gray-200 bg-white px-4 py-3 sm:px-6">
      <div className="flex flex-1 justify-between sm:hidden">
        <button
          aria-label="Previous Page"
          disabled={currentPage === 1}
          onClick={() => handlePageChange(currentPage - 1)}
          className="relative inline-flex items-center rounded-md border border-gray-300 bg-white px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-50 disabled:bg-gray-200 disabled:text-gray-400"
        >
          Previous
        </button>
        <button
          aria-label="Next Page"
          disabled={currentPage === totalPages}
          onClick={() => handlePageChange(currentPage + 1)}
          className="relative inline-flex items-center rounded-md border border-gray-300 bg-white px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-50 disabled:bg-gray-200 disabled:text-gray-400"
        >
          Next
        </button>
      </div>

      <div className="hidden sm:flex sm:flex-1 sm:items-center sm:justify-between">
        <p className="text-sm text-gray-700">
          Showing{" "}
          <span className="font-medium">{(currentPage - 1) * limit + 1}</span>{" "}
          to{" "}
          <span className="font-medium">
            {Math.min(currentPage * limit, total)}
          </span>{" "}
          of <span className="font-medium">{total}</span> results
        </p>
        <nav
          aria-label="Pagination"
          className="isolate inline-flex -space-x-px rounded-md shadow-sm"
        >
          <button
            aria-label="Previous Page"
            data-testid="previous-button"
            onClick={() => handlePageChange(currentPage - 1)}
            disabled={currentPage === 1}
            className="relative inline-flex items-center rounded-l-md px-2 py-2 text-gray-400 ring-1 ring-inset ring-gray-300 hover:bg-gray-50 focus:z-20 disabled:bg-gray-100 disabled:text-gray-400"
          >
            <FaAngleLeft className="h-5 w-5" aria-hidden="true" />
          </button>
          {Array.from({ length: totalPages }, (_, index) => (
            <button
              key={index + 1}
              aria-label={`Page ${index + 1}`}
              onClick={() => handlePageChange(index + 1)}
              className={`relative inline-flex items-center px-4 py-2 text-sm font-medium ${
                currentPage === index + 1
                  ? "z-10 bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
                  : "text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
              }`}
            >
              {index + 1}
            </button>
          ))}
          <button
            aria-label="Next Page"
            data-testid="next-button"
            onClick={() => handlePageChange(currentPage + 1)}
            disabled={currentPage === totalPages}
            className="relative inline-flex items-center rounded-r-md px-2 py-2 text-gray-400 ring-1 ring-inset ring-gray-300 hover:bg-gray-50 focus:z-20 disabled:bg-gray-100 disabled:text-gray-400"
          >
            <FaAngleRight className="h-5 w-5" aria-hidden="true" />
          </button>
        </nav>
      </div>
    </div>
  )
}

export default Pagination
