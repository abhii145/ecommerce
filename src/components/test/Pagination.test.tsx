import React from "react"
import { render, fireEvent } from "@testing-library/react"
import Pagination from "../Pagination"

describe("Pagination component", () => {
  const renderPagination = (props = {}) => {
    const defaultProps = {
      currentPage: 1,
      total: 100,
      limit: 10,
      onPageChange: jest.fn(),
      ...props,
    }
    return render(<Pagination {...defaultProps} />)
  }

it("matches snapshot", () => {
  const { asFragment } = renderPagination()
  expect(asFragment()).toMatchSnapshot()
})


  it("renders Pagination component", () => {
    const { getByText } = renderPagination()
    expect(getByText(/showing/i)).toBeInTheDocument()
  })

  it("calls onPageChange when next button is clicked", () => {
    const onPageChange = jest.fn()
    const { getByText } = renderPagination({ onPageChange })
    fireEvent.click(getByText(/next/i))
    expect(onPageChange).toHaveBeenCalledWith(2)
  })

  it("calls onPageChange when previous button is clicked", () => {
    const onPageChange = jest.fn()
    const { getByText } = renderPagination({ currentPage: 2, onPageChange })
    fireEvent.click(getByText(/previous/i))
    expect(onPageChange).toHaveBeenCalledWith(1)
  })

  it("disables previous button on first page", () => {
    const { getByText } = renderPagination()
    expect(getByText(/previous/i)).toBeDisabled()
  })

  it("disables next button on last page", () => {
    const { getByText } = renderPagination({ currentPage: 10 })
    expect(getByText(/next/i)).toBeDisabled()
  })

  it("calls onPageChange when a page number is clicked", () => {
    const onPageChange = jest.fn()
    const { getByText } = renderPagination({ onPageChange })
    fireEvent.click(getByText("3"))
    expect(onPageChange).toHaveBeenCalledWith(3)
  })

  it("calls handlePageChange when previous button is clicked in mobile view", () => {
    const onPageChange = jest.fn()
    const { getByText } = renderPagination({ currentPage: 2, onPageChange })
    fireEvent.click(getByText(/previous/i))
    expect(onPageChange).toHaveBeenCalledWith(1)
  })

  it("calls handlePageChange when next button is clicked in mobile view", () => {
    const onPageChange = jest.fn()
    const { getByText } = renderPagination({ onPageChange })
    fireEvent.click(getByText(/next/i))
    expect(onPageChange).toHaveBeenCalledWith(2)
  })

  it("calls handlePageChange when previous button is clicked in desktop view", () => {
    const onPageChange = jest.fn()
    const { container } = renderPagination({ currentPage: 2, onPageChange })
    const previousButton = container.querySelector('button[data-testid="previous-button"]')
    fireEvent.click(previousButton)
    expect(onPageChange).toHaveBeenCalledWith(1)
  })

  it("calls handlePageChange when next button is clicked in desktop view", () => {
    const onPageChange = jest.fn()
    const { container } = renderPagination({ onPageChange })
    const nextButton = container.querySelector('button[data-testid="next-button"]')
    fireEvent.click(nextButton)
    expect(onPageChange).toHaveBeenCalledWith(2)
  })
})