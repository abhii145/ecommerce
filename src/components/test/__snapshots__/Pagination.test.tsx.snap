// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-between border-t border-gray-200 bg-white px-4 py-3 sm:px-6"
  >
    <div
      class="flex flex-1 justify-between sm:hidden"
    >
      <button
        aria-label="Previous Page"
        class="relative inline-flex items-center rounded-md border border-gray-300 bg-white px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-50 disabled:bg-gray-200 disabled:text-gray-400"
        disabled=""
      >
        Previous
      </button>
      <button
        aria-label="Next Page"
        class="relative inline-flex items-center rounded-md border border-gray-300 bg-white px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-50 disabled:bg-gray-200 disabled:text-gray-400"
      >
        Next
      </button>
    </div>
    <div
      class="hidden sm:flex sm:flex-1 sm:items-center sm:justify-between"
    >
      <p
        class="text-sm text-gray-700"
      >
        Showing 
        <span
          class="font-medium"
        >
          1
        </span>
         to 
        <span
          class="font-medium"
        >
          10
        </span>
         of 
        <span
          class="font-medium"
        >
          100
        </span>
         results
      </p>
      <nav
        aria-label="Pagination"
        class="isolate inline-flex -space-x-px rounded-md shadow-sm"
      >
        <button
          aria-label="Previous Page"
          class="relative inline-flex items-center rounded-l-md px-2 py-2 text-gray-400 ring-1 ring-inset ring-gray-300 hover:bg-gray-50 focus:z-20 disabled:bg-gray-100 disabled:text-gray-400"
          data-testid="previous-button"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 256 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
            />
          </svg>
        </button>
        <button
          aria-label="Page 1"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium z-10 bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
        >
          1
        </button>
        <button
          aria-label="Page 2"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        >
          2
        </button>
        <button
          aria-label="Page 3"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        >
          3
        </button>
        <button
          aria-label="Page 4"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        >
          4
        </button>
        <button
          aria-label="Page 5"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        >
          5
        </button>
        <button
          aria-label="Page 6"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        >
          6
        </button>
        <button
          aria-label="Page 7"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        >
          7
        </button>
        <button
          aria-label="Page 8"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        >
          8
        </button>
        <button
          aria-label="Page 9"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        >
          9
        </button>
        <button
          aria-label="Page 10"
          class="relative inline-flex items-center px-4 py-2 text-sm font-medium text-gray-900 ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
        >
          10
        </button>
        <button
          aria-label="Next Page"
          class="relative inline-flex items-center rounded-r-md px-2 py-2 text-gray-400 ring-1 ring-inset ring-gray-300 hover:bg-gray-50 focus:z-20 disabled:bg-gray-100 disabled:text-gray-400"
          data-testid="next-button"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 256 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            />
          </svg>
        </button>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;
